cognite/client/data_classes/assets.py

```diff
@@ -449,3 +471,292 @@ def dump(self, camel_case: bool = False) -> Dict[str, Any]:
+        return self.validate(verbose=True, output_file=output_file, on_error="ignore")
+
+    def groupby_parent_xid(self) -> Dict[Optional[str], List[Asset]]:
+        """Returns a mapping from parent external ID to a list of its direct children.
+
+        Note:
+            If the AssetHierarchy was initialized with `ignore_orphans=True`, all orphans assets,
+            if any, are returned as part of the root assets in the mapping and can be accessed by
+            `mapping[None]`, since root assets by definition have parent external ID = None.
+        """
+        self.is_valid(on_error="raise")
+        # Sort (on parent) as required by groupby. We use hash to avoid comparing string with None:
+        parent_sorted = sorted(self._assets, key=lambda a: hash(self._get_parent(a)))
+        mapping = {
+            parent: list(child_assets)
+            for parent, child_assets in itertools.groupby(parent_sorted, key=self._get_parent)
+        }
+        if not (self._ignore_orphans and self._orphans):
+            return mapping
+
+        mapping.setdefault(None, [])
+        mapping[None].extend(
+            itertools.chain.from_iterable(
+                mapping.pop(parent, []) for parent in set(map(self._get_parent, self._orphans))
+            )
+        )
+        return mapping
+
+    def _inspect_attributes(self) -> Tuple[List[Asset], List[Asset], List[Asset], List[Asset], Dict[str, List[Asset]]]:
+        invalid, orphans, roots, unsure_parents, duplicates = [], [], [], [], defaultdict(list)
+        xid_count = Counter(map(self._get_xid, self._assets))
+
+        for asset in self._assets:
+            id_, xid, name = asset.id, asset.external_id, asset.name
+            if xid is None or name is None or len(name) < 1 or id_ is not None:
+                invalid.append(asset)
+                continue  # Don't report invalid as part of any other group
+
+            if xid_count[xid] > 1:
+                duplicates[xid].append(asset)
+
+            if (pxid := asset.parent_external_id) is (pid := asset.parent_id) is None:
+                roots.append(asset)
+            elif pxid is not None and pid is not None:
+                unsure_parents.append(asset)  # Both parent links are given
+            elif pxid not in xid_count:
+                orphans.append(asset)  # Only parent XID given, but not part of assets given
+
+        return roots, orphans, invalid, unsure_parents, dict(duplicates)
+
+    def _preconditions_for_cycle_check_are_met(self, on_error: OnErrorOptions) -> bool:
+        if not (
+            self._invalid or self._unsure_parents or self._duplicates or (self._orphans and not self._ignore_orphans)
+        ):
+            return True
+        self._on_error(on_error, "Unable to run cycle-check before basic issues are fixed.")
+        return False
+
+    def _locate_cycles(self) -> Tuple[int, List[List[str]]]:
+        edges = dict(map(op.attrgetter("external_id", "parent_external_id"), self._assets))
```

Comment:
@haakonvt 
isn't a dict-comprehension not easier -- more pythonic -- to read than this func-style?
```
edges = {asset["external_id"] : asset["parent_external_id"] for asset in self._assets}
```

Reply:
Haha jajajaja I agree 😉 I'm pretty sure it was faster in the older days of Python, but that does not seem to be the case anymore (3.9.16 shows equal performance). Btw, I guess you meant:
```py
edges = {asset.external_id: asset.parent_external_id for asset in self._assets}
```